### PR TITLE
Feature/pulsar

### PR DESCRIPTION
**Pull request by h-guo18**
## Description

加入搜索引擎和消息队列，实现搜索功能；

backend将洞、帖、回复的增删改信息通过pulsar传递给typesense，由后者记录在其数据库中并向backend提供搜索接口；

消息队列：pulsar，搜索引擎：typesense；

- [ ] 功能：在typesense中记录洞、帖、回复的增删改
- [ ] 功能：typesense-backend接口，返回搜索结果给backend

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] Code follows the code style of this project.
- [x] Changes follow the **CONTRIBUTING** guidelines.
- [x] Update necessary documentation accordingly.
- [ ] Lint and tests pass locally with the changes.
- [ ] Check issues and pull requests first. You don't want to duplicate effort.

## Other information

### CODE REVIEW

File: backend/src/bin/consumer/main.rs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+//         Err(e) => panic!("Error initial logger: {}", e),
+//     }
+// }
+
+async fn create_typesense_collections() -> Result<(), reqwest::Error> {
+    //create typesense collections
+    let collection_burrows = json!({
+      "name": "burrows",
+      "fields": [
+        {"name": "id", "type": "int64"},
```
Comment (h-guo18) [inline backend/src/bin/consumer/main.rs:38]: 改变量名

File: backend/src/bin/consumer/main.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    });
+    let collection_posts = json!({
+      "name": "posts",
+      "fields": [
+        {"name": "id", "type": "int64" },
+        {"name": "title", "type": "string" },
+        {"name": "burrow_id", "type": "int64" },
+        {"name": "created_time", "type": "string"},
+        {"name": "last_modified_time", "type": "string"},
+        {"name": "post_type", "type": "int32"},
```
Comment (h-guo18) [inline backend/src/bin/consumer/main.rs:53]: int16，如果支持

File: backend/src/bin/consumer/main.rs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        {"name": "id", "type": "int32" },
+        {"name": "post_id", "type": "int64"},
+        {"name": "created_time", "type": "string"},
+        {"name": "content", "type": "string"}
+      ]
+    });
+    let client = reqwest::Client::new();
+    for each in [collection_burrows, collection_posts, collection_replies].iter() {
+        let res = client.build_post("/collections").json(&each).send().await?;
+        // TODO: match the status code of Response here, to see whether it is successfully created or is already created, or failed
```
Comment (h-guo18) [inline backend/src/bin/consumer/main.rs:69]: 匹配错误

File: backend/src/bin/consumer/main.rs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    Ok(())
+}
+
+pub trait Typesense {
+    fn build_post(&self, uri: &str) -> reqwest::RequestBuilder;
+    fn build_delete(&self, uri: &str) -> reqwest::RequestBuilder;
+    fn build_patch(&self, uri: &str) -> reqwest::RequestBuilder;
+}
+
+impl Typesense for reqwest::Client {
```
Comment (h-guo18) [inline backend/src/bin/consumer/main.rs:80]: 如果需要加新接口

File: backend/src/bin/consumer/main.rs [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                    "id":data.data["id"],
+                    "title":data.data["title"],
+                    "introduction":data.data["introduction"],
+                    "created_time":data.operation_time,
+                    "last_modified_time":data.operation_time
+                });
+
+                match client
+                    .build_post("/collections/burrows/documents")
+                    .body(serde_json::to_string(&operation).unwrap())
```
Comment (h-guo18) [inline backend/src/bin/consumer/main.rs:168]: 可以传结构体

File: backend/src/bin/consumer/main.rs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        //     break;
+        // }
+        match (data.operation_type, data.operation_level) {
+            (OperationType::New, OperationLevel::Burrow) => {
+                // TODO: define a struct here, not using direct json
+                let operation = json!({
+                    "id":data.data["id"],
+                    "title":data.data["title"],
+                    "introduction":data.data["introduction"],
+                    "created_time":data.operation_time,
```
Comment (h-guo18) [inline backend/src/bin/consumer/main.rs:162]: 删掉

File: backend/src/bin/consumer/main.rs (hunk 2 above)
Comment (h-guo18) [reply backend/src/bin/consumer/main.rs:53]: tag+section

File: backend/src/bin/consumer/main.rs (hunk 1 above)
Comment (h-guo18) [reply backend/src/bin/consumer/main.rs:38]: 真实id：uid，洞：burrow_id

File: backend/src/bin/consumer/main.rs (hunk 2 above)
Comment (h-guo18) [reply backend/src/bin/consumer/main.rs:53]: tags：分区+剩余tag，sec是分区

File: backend/src/bin/consumer/main.rs [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        // TODO: match the status code of Response here, to see whether it is successfully created or is already created, or failed
+        match client.build_post("/collections").json(&each).send().await {
+            Ok(a) => match a.status().as_u16() {
+                201 => (),
+                400 => panic!(
+                    "Bad Request - The request could not be understood due to malformed syntax."
+                ),
+                401 => panic!("Unauthorized - Your API key is wrong."),
+                404 => panic!("Not Found - The requested resource is not found."),
+                409 => panic!("Conflict - When a resource already exists."),
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:78]: This should not panic. Imagine a case: you run this script twice, so you create the resource twice. This should not stop the program running: You have it created correctly!

File: backend/src/bin/consumer/main.rs [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+                    Ok(r) => println!("add new reply.{:?}", r),
+                    Err(e) => println!("add new reply failed {:?}", e),
+                }
+            }
+            PulsarSearchData::UpdateBurrow(burrow) => {
+                // TODO: read from typesense first, check the time if it is newer than the one in typesense
+                let search_result:SearchResult = serde_json::from_str(&client
+                .build_get(&format!("/collections/burrows/documents/search?q={}&query_by=burrow_id&filter_by=&sort_by=",burrow.burrow_id))
+                .send()
+                .await.unwrap().text().await.unwrap()).unwrap();
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:228]: do not use so many times of unwrap here. It one of them panic, you will never know who causes it.

File: backend/src/bin/consumer/main.rs [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+                match search_result.found {
+                    0 => println!("Burrow to update does not exist!"),
+                    _ => {
+                        let present_burrow: TypesenseBurrowData =
+                            serde_json::from_value(search_result.hits[0].clone()).unwrap();
+                        match DateTimeWithTimeZone::parse_from_rfc3339(&present_burrow.update_time)
+                            .unwrap()
+                            .timestamp()
+                            < burrow.update_time.timestamp()
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:237]: Can it be directly compared? Not tranforming to timestamp?

File: backend/src/bin/consumer/main.rs [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+                    .await
+                {
+                    Ok(r) => println!("add new reply.{:?}", r),
+                    Err(e) => println!("add new reply failed {:?}", e),
+                }
+            }
+            PulsarSearchData::UpdateBurrow(burrow) => {
+                // TODO: read from typesense first, check the time if it is newer than the one in typesense
+                let search_result:SearchResult = serde_json::from_str(&client
+                .build_get(&format!("/collections/burrows/documents/search?q={}&query_by=burrow_id&filter_by=&sort_by=",burrow.burrow_id))
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:226]: here you have the burrow id, I think you don't need to perform a search. Simply get the document.

File: backend/src/routes/search.rs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+}
+///Search keyword in post title/replies. Result filter by tag.
+#[get("/post/<keyword>/<tag>")]
+async fn search_post_keyword(
+    searchpool: Connection<TypesenseSearch>,
+    keyword: &str,
+    tag: &str,
+) -> String {
+    let client = searchpool.into_inner();
+    let uri = format!("/collections/replies/documents/search?q={}&query_by=content&filter_by=tag:= {}&group_by=post_id&sort_by=update_time", keyword,tag);
```
Comment (BobAnkh) [inline backend/src/routes/search.rs:66]: Can this interface work?I see here a space inside the uri

File: backend/src/routes/search.rs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        Err(e) => return format!("build_get send Error: {:?}", e),
+    };
+    let result: SearchResult = serde_json::from_str(&response).unwrap();
+    match result.found {
+        0 => "No results!".to_string(),
+        _ => json!({
+            "found":result.found,
+            "hits":result.hits
+        })
+        .to_string(),
```
Comment (BobAnkh) [inline backend/src/routes/search.rs:36]: return json, not string

File: backend/src/routes/search.rs [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    }
```
Comment (BobAnkh) [inline backend/src/routes/search.rs:101]: when given only postid, just use sea-orm of postgres, no need to touch typesense.

File: backend/src/routes/search.rs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
```
Comment (BobAnkh) [inline backend/src/routes/search.rs:56]: when given only burrowid, use sea-orm to search postgres, no need to touch typesense

**Review by BobAnkh — CHANGES_REQUESTED** [review]

File: backend/src/bin/consumer/main.rs [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+                            }
+                            false => println!("The burrow is already up to date."),
+                        }
+                    }
+                }
+            }
+            PulsarSearchData::UpdatePost(post) => {
+                // TODO: read from typesense first, check the time if it is newer than the one in typesense
+                let search_result:SearchResult = serde_json::from_str(&client
+                    .build_get(&format!("/collections/posts/documents/search?q={}&query_by=post_id&filter_by=&sort_by=",post.post_id))
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:270]: retrieve it when you have the post id

File: backend/src/bin/consumer/main.rs [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+                    .await.unwrap().text().await.unwrap()).unwrap();
+                match search_result.found {
+                    0 => println!("Post to update does not exist!"),
+                    _ => {
+                        let present_post: TypesenseBurrowData =
+                            serde_json::from_value(search_result.hits[0].clone()).unwrap();
+                        match DateTimeWithTimeZone::parse_from_rfc3339(&present_post.update_time)
+                            .unwrap()
+                            .timestamp()
+                            < post.update_time.timestamp()
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:281]: same problem as above mentioned

File: backend/src/bin/consumer/main.rs [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+                            }
+                            false => println!("The burrow is already up to date."),
+                        }
+                    }
+                }
+            }
+            PulsarSearchData::UpdateReply(reply) => {
+                // TODO: read from typesense first, check the time if it is newer than the one in typesense
+                let search_result:SearchResult = serde_json::from_str(&client
+                    .build_get(&format!("/collections/replies/documents/search?q={}&query_by=reply_id&filter_by=&sort_by=",reply.reply_id))
```
Comment (BobAnkh) [inline backend/src/bin/consumer/main.rs:305]: same problem

File: backend/src/req/pulsar.rs [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+#[derive(Serialize, Deserialize)]
+pub struct SearchRequest {
+    pub keyword: Option<String>,
+    pub id: Option<i64>,
+    pub tag: Option<String>,
+    pub order: Option<String>,
+    pub area: Option<String>,
+    pub page: Option<i16>,
+    pub filter_by:Option<String>
+}
```
Comment (BobAnkh) [inline backend/src/req/pulsar.rs:315]: Use enum to filter different kind of search

File: backend/src/routes/search.rs [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+}
+///redirect to other functions.
+#[post("/search", data = "<data>", format = "json")]
+async fn search(data: Json<SearchRequest>) -> Redirect {
+    match &data.area {
+        Some(area) => match &area[..] {
+            "burrow" => match (data.keyword.as_ref(), data.id, data.tag.as_ref()) {
+                (Some(_keyword), None, None) => Redirect::temporary(uri!(search_burrow_keyword())),
+                (None, Some(id), None) => Redirect::temporary(uri!(retrieve_burrow(id))),
+                (None, None, Some(_tag)) => {
```
Comment (BobAnkh) [inline backend/src/routes/search.rs:31]: Do not use redirect.

File: backend/src/routes/search.rs [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+pub async fn init(rocket: Rocket<Build>) -> Rocket<Build> {
+    rocket.mount(
+        "/search",
+        routes![
+            search,
+            search_burrow_keyword,
+            retrieve_burrow,
+            search_post_keyword,
+            search_post_tag,
+            retrieve_post
```
Comment (BobAnkh) [inline backend/src/routes/search.rs:19]: No need to have so much function and interfaces. merge them into one

**Review by BobAnkh — CHANGES_REQUESTED** [review]
Change your search request to enum. I have mentioned this.

Merge all the search interfaces into one. Use enum to decide the search type.

**Comment by h-guo18** [comment]
image
image
image
image
image
image
image
image
image
image

**Pull request closed without merging**